### PR TITLE
[#15] [Integrate] As a user, I can keep the logged in state whenever reopening the application

### LINKED ISSUE #15: [Integrate] As a user, I can keep the logged in state whenever reopening the application
## Why
After the user sign in, he can keep logged in state when revisiting the application

## Acceptance Criteria
- When the user closes the application and reopens it again, he doesn't need to enter the email and the password to sign in. Instead, the website will redirect the user to the home page automatically.

### PR DESCRIPTION
**Pull request by manh-t**
Close #15 

## What happened 👀

- After the user signs in, he can come to the Dashboard screen directly the next time he opens the website.
- After logged in, the user will be redirect to Dashboard screen if he tries to navigate to Sign In screen

## Insight 📝

Small refactor:
- Rename `signIn` to `signInAsyncThunk` to distinguish between `sign in` from adapter and `sign in` from redux.

## Proof Of Work 📹

https://github.com/manh-t/react-survey/assets/60863885/1b6f1d2c-4523-47c8-acbd-a89563d5df0c

### CODE REVIEW

**Comment by manh-t** [comment]
🚀 Deployed on https://64c1e992af01d4409307dbb9--manh-react-survey.netlify.app

**Comment by manh-t** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>Congrats, coverage is good</td>
    </tr>
  </tbody>
</table>

## Coverage in All Files
File | Line Coverage | Statement Coverage | Function Coverage | Branch Coverage
---- | ------------: | -----------------: | ----------------: | --------------:
[src/../Dashboard/index.tsx](../blob/d6f74f7/src/screens/Dashboard/index.tsx) | (1/6) 17% | (1/6) 17% | (0/4) 0% | (0/0) NaN%
[src/../v1/requestManager.ts](../blob/d6f74f7/src/lib/request/v1/requestManager.ts) | (9/18) 50% | (9/18) 50% | (4/6) 67% | (1/3) 33%
[src/config/index.ts](../blob/d6f74f7/src/config/index.ts) | (5/9) 56% | (5/9) 56% | (0/3) 0% | (0/4) 0%
[src/tests/error.ts](../blob/d6f74f7/src/tests/error.ts) | (2/3) 67% | (2/3) 67% | (1/2) 50% | (1/3) 33%
[src/screens/SignIn/index.tsx](../blob/d6f74f7/src/screens/SignIn/index.tsx) | (12/16) 75% | (12/16) 75% | (3/6) 50% | (8/8) 100%
[src/../Content/index.tsx](../blob/d6f74f7/src/components/Dashboard/Content/index.tsx) | (8/10) 80% | (8/10) 80% | (4/6) 67% | (3/5) 60%
[src/../index.tsx](../blob/d6f74f7/src/routes/ProtectedRoute/index.tsx) | (4/5) 80% | (4/5) 80% | (2/2) 100% | (1/2) 50%
[src/App.tsx](../blob/d6f74f7/src/App.tsx) | (3/3) 100% | (3/3) 100% | (1/1) 100% | (0/0) NaN%
[src/index.tsx](../blob/d6f74f7/src/index.tsx) | (3/3) 100% | (3/3) 100% | (0/0) NaN% | (0/0) NaN%
[src/../Authentication/index.ts](../blob/d6f74f7/src/adapters/Authentication/index.ts) | (2/2) 100% | (2/2) 100% | (1/1) 100% | (0/0) NaN%
[src/adapters/Base/index.ts](../blob/d6f74f7/src/adapters/Base/index.ts) | (8/8) 100% | (8/8) 100% | (2/2) 100% | (2/2) 100%
[src/adapters/BaseAuth/index.ts](../blob/d6f74f7/src/adapters/BaseAuth/index.ts) | (10/10) 100% | (10/10) 100% | (4/4) 100% | (2/2) 100%
[src/adapters/Survey/index.ts](../blob/d6f74f7/src/adapters/Survey/index.ts) | (1/1) 100% | (1/1) 100% | (1/1) 100% | (0/0) NaN%
[src/components/Alert/index.tsx](../blob/d6f74f7/src/components/Alert/index.tsx) | (3/3) 100% | (3/3) 100% | (2/2) 100% | (0/0) NaN%
[src/../index.tsx](../blob/d6f74f7/src/components/BackgroundImage/index.tsx) | (3/3) 100% | (3/3) 100% | (1/1) 100% | (2/2) 100%
Other (22 more) | (98/98) 100% | (98/98) 100% | (34/34) 100% | (28/29) 97%
Total | (172/198) 87% | (172/198) 87% | (60/75) 80% | (48/60) 80%

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 3aadb5c4ed29ded049fb3ae88f102051848a7681
</p>

File: src/screens/Dashboard/index.test.tsx [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    (useAppSelector as jest.Mock).mockImplementation((callback) => callback(mockState));
+    (useAppDispatch as jest.Mock).mockImplementation(() => mockDispatch);
+  });
+
+  afterEach(() => {
+    jest.clearAllMocks();
+  });
+
+  describe('given the token is undefined', () => {
+    it('navigates to the SignIn screen', () => {
```
Comment (hoangmirs) [inline src/screens/Dashboard/index.test.tsx]: We can skip this test as it doesn't have any assertions
```suggestion
    xit('navigates to the SignIn screen', () => {
```

File: src/screens/Dashboard/index.test.tsx [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+      render(<TestComponent />);
+
+      // TODO replace by other unittest later
```
Comment (hoangmirs) [inline src/screens/Dashboard/index.test.tsx]:
```suggestion
      // TODO: replace by another unit test later
```

File: src/store/reducers/Authentication/index.ts [hunk 3]
```diff
@@ -1,36 +1,43 @@
...
 }
 
 export const initialState: AuthenticationState = {
   loading: false,
   success: false,
+  token: token,
 };
 
-export const signIn = createAsyncThunk('auth/signIn', signInAsync);
+export const signInAsyncThunk = createAsyncThunk('auth/signIn', signInThunkCreator);
```
Comment (hoangmirs) [inline src/store/reducers/Authentication/index.ts]: I think we can keep the constant name `signIn` as this is just a normal action. We rarely name the action like that.
https://redux-toolkit.js.org/api/createAsyncThunk

Comment (manh-t) [reply src/store/reducers/Authentication/index.ts]: @hoangmirs I have a problem when writing the unit test that `signIn` from adapter has the same name 🙈 But it's okay, let me revert it back and add the alias if there is the duplicated name

File: src/routes/ProtectedRoute/index.tsx [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+import { Navigate, Outlet } from 'react-router-dom';
+
+import { useAppSelector } from 'hooks';
+import { paths } from 'routes';
+
+// Tutorial Link:
+// https://blog.logrocket.com/complete-guide-authentication-with-react-router-v6/
+// https://blog.logrocket.com/handling-user-authentication-redux-toolkit/
+const ProtectedRoute = (): JSX.Element => {
+  const { token } = useAppSelector((state) => state.auth);
```
Comment (nvminhtue) [inline src/routes/ProtectedRoute/index.tsx:11]: Is it always having `token` (regardless the value) in the `state.auth` object? It might crash the app if `state.auth` returns undefined 🤔 Be careful with that

File: src/screens/Dashboard/index.test.tsx [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+import DashBoardScreen from '.';
+
+const mockUseNavigate = jest.fn();
+const mockDispatch = jest.fn();
+
+jest.mock('hooks');
+
+jest.mock('react-router-dom', () => ({
+  ...(jest.requireActual('react-router-dom') as jest.Mock),
```
Comment (nvminhtue) [inline src/screens/Dashboard/index.test.tsx]: Do we need this 🤔

File: src/store/reducers/Authentication/index.test.ts [hunk 6]
```diff
@@ -88,7 +86,7 @@ describe('auth slice', () => {
         expect(errorAction.payload).toEqual({ data: mockError.response?.data, status: mockError.response?.status });
         expect(errorAction.meta.arg).toBe(input);
 
-        expect(dispatch).toHaveBeenNthCalledWith(1, signIn.pending(errorAction.meta.requestId, input));
+        expect(dispatch).toHaveBeenNthCalledWith(1, signInAsyncThunk.pending(errorAction.meta.requestId, input));
```
Comment (nvminhtue) [inline src/store/reducers/Authentication/index.test.ts]: Why don't you use await or Promise resolve in the function and assert the result (returned from success/failed cases) 🤔. That's meaningless if we just checking the function is called in the pending regardless checking the final results

File: src/routes/ProtectedRoute/index.tsx (hunk 4 above)
Comment (hoangmirs) [reply src/routes/ProtectedRoute/index.tsx:11]: It seems it was added in the initialState [here](https://github.com/manh-t/react-survey/pull/56/files#diff-445a58788e0d28fcf32135554f18f6bcf0b7d194e74cdf6edd7d0855073e5e94R21)

File: src/store/reducers/Authentication/index.test.ts (hunk 6 above)
Comment (manh-t) [reply src/store/reducers/Authentication/index.test.ts]: @nvminhtue ah I got it. It is unnecessary to verify the call here. Let me verify the result instead.

File: src/screens/Dashboard/index.test.tsx (hunk 5 above)
Comment (manh-t) [reply src/screens/Dashboard/index.test.tsx]: @nvminhtue I think it's redundant. Let me check 🙆‍♂️

File: src/routes/ProtectedRoute/index.tsx (hunk 4 above)
Comment (manh-t) [reply src/routes/ProtectedRoute/index.tsx:11]: @nvminhtue don't worry, we always have an initial state 🙋

File: src/store/reducers/Authentication/index.test.ts (hunk 6 above)
Comment (manh-t) [reply src/store/reducers/Authentication/index.test.ts]: @nvminhtue I updated the verify's part in [3aadb5c](https://github.com/manh-t/react-survey/pull/56/commits/3aadb5c4ed29ded049fb3ae88f102051848a7681)

File: src/screens/Dashboard/index.test.tsx (hunk 5 above)
Comment (manh-t) [reply src/screens/Dashboard/index.test.tsx]: @nvminhtue removed in [3aadb5c](https://github.com/manh-t/react-survey/pull/56/commits/3aadb5c4ed29ded049fb3ae88f102051848a7681)

File: src/screens/Dashboard/index.test.tsx (hunk 2 above)
Comment (manh-t) [reply src/screens/Dashboard/index.test.tsx]: @hoangmirs the linter suggested me updating `xit` to `it.skip` instead. I udpated it in [3aadb5c](https://github.com/manh-t/react-survey/pull/56/commits/3aadb5c4ed29ded049fb3ae88f102051848a7681)

File: src/routes/index.tsx [hunk 7]
```diff
@@ -4,20 +4,26 @@ import { RouteObject } from 'react-router-dom';
...
+      {
+        path: paths.root,
+        element: <DashBoardScreen />,
+      },
+    ],
   },
   {
-    path: paths.dashboard,
-    element: <DashBoardScreen />,
+    path: paths.signIn,
```
Comment (nvminhtue) [inline src/routes/index.tsx:25]: Do we need the test here? Maybe the user is not logged in and the path is `paths.signIn`?

**Review by tyrro — APPROVED** [review]

**Review by hoangmirs — APPROVED** [review]

File: src/routes/index.tsx (hunk 7 above)
Comment (manh-t) [reply src/routes/index.tsx:25]: @nvminhtue I'm having trouble finding ways to test this file. As a solution, I've written an integration test to confirm the behavior of "the user is not logged in and the path is paths.signIn" in [this PR](https://github.com/manh-t/react-survey/pull/62/files#diff-d299c7915109b3ddc4015b3d2427383bb98b029e9d81069c3fb2e2975a0c3baeR34). Do you think this is a sufficient way to test it, or do you have any other suggestions? :pray:

Comment (nvminhtue) [reply src/routes/index.tsx:25]: I see, but the test in the mentioned PR doesn't cover the `paths.signIn` assertion like the way `paths.dashboard` did. But as long as you cover that case, it's okay to me, just placing tests for a single feature in different PRs is hard to follow though.

**Review by nvminhtue — APPROVED** [review]

File: src/routes/index.tsx (hunk 7 above)
Comment (manh-t) [reply src/routes/index.tsx:25]: @nvminhtue yeah, that's why I ask if you have any suggestions for testing this file 😄 Otherwise, I will try this case in the integration test. Anw, thank you 🙏
